components/tidb_query_datatype/src/def/field_type.rs

```diff
@@ -280,6 +280,16 @@ pub trait FieldTypeAccessor {
     fn is_bool(&self) -> bool {
         self.flag().contains(FieldTypeFlag::IS_BOOLEAN)
     }
+
+    #[inline]
+    fn need_restored_data(&self) -> bool {
+        self.is_non_binary_string_like()
+            && (!self
+                .collation()
+                .map(|col| col == Collation::Utf8Mb4Bin)
```

Comment:
We may simplify this condition to `!self.collation() == Some(Collation::Utf8Mb4Bin)`